File: cgol_c/grid.c
#include <stdlib.h>
#include <stdio.h>
#include "grid.h"

grid grid_init(int rows, int cols) {
    int **data = (int**)malloc(rows * sizeof(int*));
    for (int i=0; i<rows; i++) {
        *(data+i) = (int*)malloc(cols * sizeof(int));
    }
    for(int i=0; i<rows; i++) {
        for (int j=0; j<cols; j++) {
            *(*(data+i)+j) = 0;
        }
    }
    grid ret_grid;
    ret_grid.data = data;
    ret_grid.rows = rows;
    ret_grid.cols = cols;
    return ret_grid;
}

void grid_free(grid* matrix) {
    for (int i=0; i<matrix->rows; i++) {
        free(matrix->data[i]);
    }
    free(matrix->data);
}

int grid_set(grid* matrix, int x, int y) {
    if (x >= matrix->rows - 1 || y >= matrix->cols - 1) {
        return 1;
    }
    matrix->data[x][y] = 1;
    return 0;
}

void grid_show(grid* matrix) {
    for (int i=0; i<matrix->rows; i++) {
        for (int j=0; j<matrix->cols; j++) {
            if (matrix->data[i][j] == 0) {
                printf(" ");
            } else {
                printf("#");
            }
        }
        printf("\n");
    }
}

int grid_num_alive(grid* matrix) {
    int num_alive = 0;
    for (int i=0; i<matrix->rows; i++) {
        for (int j=0; j<matrix->cols; j++) {
            if (matrix->data[i][j] == 1) {
                num_alive ++;
            }
        }
    }
    return num_alive;
}

int grid_state(int*  state, int* alive) {
    if (*state == 1) {
        if (*alive < 2) {
            return 0;
        } else if (*alive == 2 || *alive == 3) {
            return 1;
        } else {
            return 0;
        }
    } else {
        if (*alive == 3) {
            return 1;
        } else {
            return *state;
        }
    }
}

int grid_next(grid* matrix) {
    grid new_grid = grid_init(matrix->rows, matrix->cols);
    for (int i=0; i<matrix->rows; i++) {
        for (int j=0; j<matrix->cols; j++) {
            //check for neighbours
            if (i == 0) {
                //check for top row
                if (j == 0) {
                    //check top left element
                    int alive = 0;
                    if (matrix->data[i][j+1] == 1) { alive++; }
                    if (matrix->data[i+1][j] == 1) { alive ++; }
                    if (matrix->data[i+1][j+1] == 1) { alive ++; }
                    int new_state = grid_state(&matrix->data[i][j], &alive);
                    new_grid.data[i][j] = new_state;
                } else if (j == matrix->cols - 1) {
                    //check for top right element
                    int alive = 0;
                    if (matrix->data[i][j-1] == 1) { alive++; }
                    if (matrix->data[i+1][j] == 1) { alive ++; }
                    if (matrix->data[i+1][j-1] == 1) { alive ++; }
                    int new_state = grid_state(&matrix->data[i][j], &alive);
                    new_grid.data[i][j] = new_state;
                } else {
                    //check for top middle elements
                    int alive = 0;
                    if (matrix->data[i][j-1] == 1) { alive++; }
                    if (matrix->data[i][j+1] == 1) { alive++; }
                    if (matrix->data[i+1][j-1] == 1) { alive++; }
                    if (matrix->data[i+1][j] == 1) { alive++; }
                    if (matrix->data[i+1][j+1] == 1) { alive++; }
                    int new_state = grid_state(&matrix->data[i][j], &alive);
                    new_grid.data[i][j] = new_state;
                }
            } else if (i == matrix->rows - 1) {
                //check for bottom row
                if (j == 0) {
                    //check for bottom left element
                    int alive = 0;
                    if (matrix->data[i][j+1] == 1) { alive++; }
                    if (matrix->data[i-1][j] == 1) { alive++; }
                    if (matrix->data[i-1][j+1] == 1) { alive++; }
                    int new_state = grid_state(&matrix->data[i][j], &alive);
                    new_grid.data[i][j] = new_state;
                } else if (j == matrix->cols -1) {
                    //check for bottom right element
                    int alive = 0;
                    if (matrix->data[i][j-1] == 1) { alive++; }
                    if (matrix->data[i-1][j] == 1) { alive++; }
                    if (matrix->data[i-1][j-1] == 1) { alive++; }
                    int new_state = grid_state(&matrix->data[i][j], &alive);
                    new_grid.data[i][j] = new_state;
                } else {
                    //check for bottom middle elements
                    int alive = 0;
                    if (matrix->data[i][j+1] == 1) { alive++; }
                    if (matrix->data[i][j-1] == 1) { alive++; }
                    if (matrix->data[i-1][j-1] == 1) { alive++; }
                    if (matrix->data[i-1][j] == 1) { alive++; }
                    if (matrix->data[i-1][j+1] == 1) { alive++; }
                    int new_state = grid_state(&matrix->data[i][j], &alive);
                    new_grid.data[i][j] = new_state;
                }
            } else {
                //checking for middle rows
                if (j == 0) {
                    //check for middle left element
                    int alive = 0;
                    if (matrix->data[i][j+1] == 1) { alive++; }
                    if (matrix->data[i-1][j] == 1) { alive++; }
                    if (matrix->data[i-1][j+1] == 1) { alive++; }
                    if (matrix->data[i+1][j] == 1) { alive++; }
                    if (matrix->data[i+1][j+1] == 1) { alive++; }
                    int new_state = grid_state(&matrix->data[i][j], &alive);
                    new_grid.data[i][j] = new_state;
                } else if (j == matrix->cols - 1) {
                    //check for middle right element
                    int alive = 0;
                    if (matrix->data[i][j-1] == 1) { alive++; }
                    if (matrix->data[i-1][j] == 1) { alive++; }
                    if (matrix->data[i-1][j-1] == 1) { alive++; }
                    if (matrix->data[i+1][j] == 1) { alive++; }
                    if (matrix->data[i+1][j-1] == 1) { alive++; }
                    int new_state = grid_state(&matrix->data[i][j], &alive);
                    new_grid.data[i][j] = new_state;
                } else {
                    //check for middle middle elements
                    int alive = 0;
                    if (matrix->data[i][j+1] == 1) { alive++; }
                    if (matrix->data[i][j-1] == 1) { alive++; }
                    if (matrix->data[i-1][j-1] == 1) { alive++; }
                    if (matrix->data[i-1][j] == 1) { alive++; }
                    if (matrix->data[i-1][j+1] == 1) { alive++; }
                    if (matrix->data[i+1][j-1] == 1) { alive++; }
                    if (matrix->data[i+1][j] == 1) { alive++; }
                    if (matrix->data[i+1][j+1] == 1) { alive++; }
                    int new_state = grid_state(&matrix->data[i][j], &alive);
                    new_grid.data[i][j] = new_state;
                }
            }
        }
    }
    grid_free(matrix);
    matrix->data = new_grid.data;
    int alives = grid_num_alive(matrix);
    if (alives == 0) {
        return 1;
    } else {
        return 0;
    }
}

File: cgol_rs/src/grid.rs
use ratatui::style::{Style, Stylize};
use ratatui::widgets::Widget;

#[derive(Debug, Clone)]
pub struct Grid {
    data: Vec<Vec<bool>>,
    rows: usize,
    cols: usize,
}

impl Grid {
    pub fn new(rows: usize, cols: usize) -> Self {
        let data = vec![vec![false; cols]; rows];
        Self { data, rows, cols }
    }

    pub fn show(&self) {
        for i in 0..self.rows {
            for j in 0..self.cols {
                match self.data[i][j] {
                    true => print!("#"),
                    false => print!(" "),
                }
            }
            print!("\n")
        }
    }

    pub fn set(&mut self, indices: Vec<(usize, usize)>) {
        for item in indices {
            if item.0 > self.rows && item.1 > self.cols {
                panic!("provided indices out of bounds")
            } else {
                self.data[item.0][item.1] = true;
            }
        }
    }

    pub fn alive_count(&self) -> usize {
        let mut alive = 0;
        for i in 0..self.rows {
            for j in 0..self.cols {
                if self.data[i][j] == true {
                    alive += 1;
                }
            }
        }
        alive
    }
    
    fn get_alive(data: Vec<&bool>) -> usize {
        let mut alive = 0;
        for item in data {
            if *item == true {
                alive += 1;
            }
        }
        alive
    }

    fn get_new_state(state: &bool, alive: usize) -> bool {
        if *state == true {
            if alive < 2 {
                return false
            } else if alive == 2 || alive == 3 {
                return true
            } else {
                return false
            }
        } else {
            if alive == 3 {
                return true
            } else {
                return *state
            }
        }
    }
}

impl Iterator for Grid {
    type Item = Self;
    fn next(&mut self) -> Option<Self> {
        let mut new_grid = Self::new(self.rows, self.cols);
        for i in 0..self.rows {
            for j in 0..self.cols {
                //check for neighbours
                if i == 0 {
                    //check for top row
                    if j == 0 {
                        //check top left element
                        let alive = Self::get_alive(vec![
                            &self.data[i][j+1],
                            &self.data[i+1][j],
                            &self.data[i+1][j+1],
                        ]);
                        let new_state = Self::get_new_state(&self.data[i][j], alive);
                        new_grid.data[i][j] = new_state;
                    } else if j == self.cols - 1 {
                        //check for top right element
                        let alive = Self::get_alive(vec![
                            &self.data[i][j-1],
                            &self.data[i+1][j],
                            &self.data[i+1][j-1],
                        ]);
                        let new_state = Self::get_new_state(&self.data[i][j], alive);
                        new_grid.data[i][j] = new_state;
                    } else {
                        //check for top middle elements
                        let alive = Self::get_alive(vec![
                            &self.data[i][j-1],
                            &self.data[i][j+1],
                            &self.data[i+1][j-1],
                            &self.data[i+1][j],
                            &self.data[i+1][j+1],
                        ]);
                        let new_state = Self::get_new_state(&self.data[i][j], alive);
                        new_grid.data[i][j] = new_state;
                    }
                } else if i == self.rows - 1 {
                    //check for bottom row
                    if j == 0 {
                        //check for bottom left element
                        let alive = Self::get_alive(vec![
                            &self.data[i][j+1],
                            &self.data[i-1][j],
                            &self.data[i-1][j+1],
                        ]);
                        let new_state = Self::get_new_state(&self.data[i][j], alive);
                        new_grid.data[i][j] = new_state;
                    } else if j == self.cols -1 {
                        //check for bottom right element
                        let alive = Self::get_alive(vec![
                            &self.data[i][j-1],
                            &self.data[i-1][j],
                            &self.data[i-1][j-1],
                        ]);
                        let new_state = Self::get_new_state(&self.data[i][j], alive);
                        new_grid.data[i][j] = new_state;
                    } else {
                        //check for bottom middle elements
                        let alive = Self::get_alive(vec![
                            &self.data[i][j+1],
                            &self.data[i][j-1],
                            &self.data[i-1][j-1],
                            &self.data[i-1][j],
                            &self.data[i-1][j+1],
                        ]);
                        let new_state = Self::get_new_state(&self.data[i][j], alive);
                        new_grid.data[i][j] = new_state;
                    }
                } else {
                    //checking for middle rows
                    if j == 0 {
                        //check for middle left element
                        let alive = Self::get_alive(vec![
                            &self.data[i][j+1],
                            &self.data[i-1][j],
                            &self.data[i-1][j+1],
                            &self.data[i+1][j],
                            &self.data[i+1][j+1],
                        ]);
                        let new_state = Self::get_new_state(&self.data[i][j], alive);
                        new_grid.data[i][j] = new_state;
                    } else if j == self.cols - 1 {
                        //check for middle right element
                        let alive = Self::get_alive(vec![
                            &self.data[i][j-1],
                            &self.data[i-1][j],
                            &self.data[i-1][j-1],
                            &self.data[i+1][j],
                            &self.data[i+1][j-1],
                        ]);
                        let new_state = Self::get_new_state(&self.data[i][j], alive);
                        new_grid.data[i][j] = new_state;
                    } else {
                        //check for middle middle elements
                        let alive = Self::get_alive(vec![
                            &self.data[i][j+1],
                            &self.data[i][j-1],
                            &self.data[i-1][j-1],
                            &self.data[i-1][j],
                            &self.data[i-1][j+1],
                            &self.data[i+1][j-1],
                            &self.data[i+1][j],
                            &self.data[i+1][j+1],
                        ]);
                        let new_state = Self::get_new_state(&self.data[i][j], alive);
                        new_grid.data[i][j] = new_state;
                    }
                }
            }
        }
        *self = new_grid;       
        Some(self.clone())
    } 
}

impl Widget for &Grid {
    fn render(self, _area: ratatui::prelude::Rect, buf: &mut ratatui::prelude::Buffer) {
        for i in 0..self.rows as u16 {
            for j in 0..self.cols as u16 {
                if self.data[i as usize][j as usize] == true {
                    buf.set_string(i, j, "#", Style::new().green());
                } else {
                    buf.set_string(i, j, " ", Style::new());
                }
            }
        }
    }
}
